[V1][Metrics] Implement vllm:lora_requests_info metric

markmc opened the pull request:
```
vllm:lora_requests_info{max_lora="1",running_lora_adapters="",waiting_lora_adapters=""} 1.7395575657589855e+09
vllm:lora_requests_info{max_lora="1",running_lora_adapters="test-lora",waiting_lora_adapters=""} 1.7395575723949368e+09
vllm:lora_requests_info{max_lora="1",running_lora_adapters="test-lora",waiting_lora_adapters="test-lora"} 1.7395575717647147e+09
```

As discussed in #13303, this metric perhaps isn't the most ideal solution for the use case but, given there is an existing user, we should retain compatibility in V1 and deprecate it if we replace it with a different metric.

See also kubernetes-sigs/gateway-api-inference-extension#354

markmc committed:
[V1][Metrics] Implement vllm:lora_requests_info metric

```
vllm:lora_requests_info{max_lora="1",running_lora_adapters="",waiting_lora_adapters=""} 1.7395575657589855e+09
vllm:lora_requests_info{max_lora="1",running_lora_adapters="test-lora",waiting_lora_adapters=""} 1.7395575723949368e+09
vllm:lora_requests_info{max_lora="1",running_lora_adapters="test-lora",waiting_lora_adapters="test-lora"} 1.7395575717647147e+09
```

As discussed in #13303, this metric perhaps isn't the most ideal
solution for the use case but, given there is an existing user,
we should retain compatibility in V1 and deprecate it if we
replace it with a different metric.

See also kubernetes-sigs/gateway-api-inference-extension#354

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

robertgshaw2-redhat commented on vllm/v1/engine/output_processor.py:
```diff
@@ -53,11 +56,13 @@ def from_new_request(
         cls,
         tokenizer: AnyTokenizer,
         request: EngineCoreRequest,
+        lora_request: Optional[LoRARequest],
```
`EngineCorerRequest` already has `LoRARequest` as a member (https://github.com/vllm-project/vllm/blob/33170081f13b4e4453f0fff8df994df8d9aa6c6c/vllm/v1/engine/__init__.py#L61) so you do not need to pass the `lora_request` as an argument to this function

robertgshaw2-redhat commented on vllm/v1/engine/output_processor.py:
```diff
@@ -106,6 +112,7 @@ def abort_requests(
     def add_request(
         self,
         request: EngineCoreRequest,
+        lora_request: Optional[LoRARequest] = None,
```
same comment as above `EngineCoreRequest.lora_request` has this value already

markmc replied:
Fixed

robertgshaw2-redhat commented on vllm/v1/metrics/stats.py:
```diff
@@ -52,6 +52,12 @@ class RequestStateStats:
     last_token_ts: float = 0.0
 
 
+@dataclass
```
I recognize this is a complete nit, but I think this should go next to `SchedulerStats`. When I originally read this code, since it is between `RequestStateStats` and `FinishedRequestStats`, I thought this would be LoRA stats about a single request, when in reality it is about the global state

markmc replied:
Yep, makes sense. Fixed

robertgshaw2-redhat commented on vllm/v1/engine/output_processor.py:
```diff
@@ -241,11 +255,22 @@ def _update_stats_from_finished(self, req_state: RequestState,
             return
 
+        lora_stats: Optional[LoRAStats] = None
+        if req_state.lora_name:
```
this can just be `lora_stats = self.lora_stats.get(req_state.lora_name, None)`

robertgshaw2-redhat commented on vllm/v1/metrics/stats.py:
```diff
@@ -151,3 +168,11 @@ def update_from_finished_request(self, finish_reason: "FinishReason",
                                  decode_time=decode_time)
         self.finished_requests.append(finished_req)
+
+        if lora_stats is not None:
```
One state transition which may not be accounted for here is if a request is aborted

Aborts occur when the client disconnects from the server. The `AsyncLLM` sends an abort message to the `OutputProcessor` and to the `EngineCore`. This **may** cause the request to be finished, but I am not 100% sure that it does. This could cause the request to never be removed from `running_requests`. We should just double check this.

Additionally, if a client aborts when the request is in `lora_stats.waiting_requests` it will not ever be removed from the waiting set since the request will never be scheduled.

markmc replied:
Great point, take a look now - basically I've tried to couple this closely with `RequestState` lifecycle

robertgshaw2-redhat reviewed (changes requested):
Generally looks good. Left a small nit that will help clean up the API a bit `EngineCoreRequest` already contains `lora_request` so we do not need to pass it around as a separate arg

The one functionality issue I see is that we do not handle request aborting properly. Can connect offline about this but I left a detailed comment below.

markmc committed:
[V1][Metrics] Drop LoRA arg from RequestState.from_new_request()

The LoRA name is already available via the EngineCoreRequest arg.

PR feedback from Rob.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Align LoRAStats with SchedulerStats

Both are global state. PR feedback from Rob.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Hook up OutputProcessor._update_lora_iteration_stats()

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Fix LoRAStats lifecycle

Tie LoRAStats lifecycle more closely to RequestStats.

1. Add request to waiting_requests when we create a
   RequestStats in add_request()
2. Remove request from LoRAStats whenever we remove
   a RequestStats (both finish and abort)

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Metrics] Refactor LoRA metrics gathering

Encapsulate all of this into a LoRARequestStates class in the
v1.metrics.stats module, since its sole purpose is to gather
stats.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
> Generally looks good. Left a small nit that will help clean up the API a bit `EngineCoreRequest` already contains `lora_request` so we do not need to pass it around as a separate arg
> 
> The one functionality issue I see is that we do not handle request aborting properly. Can connect offline about this but I left a detailed comment below.

Thanks @robertgshaw2-redhat. Great feedback. This definitely warranted a bunch of cleanup. See the 5 fixup commits I pushed :+1:

robertgshaw2-redhat commented on vllm/v1/metrics/stats.py:
```diff
@@ -151,3 +166,55 @@ def update_from_finished_request(self, finish_reason: "FinishReason",
+    def __init__(self):
+        self.lora_name_to_stats: Dict[str, LoRAStats] = {}
+
+    def get_stats(self, req_state) -> Optional[LoRAStats]:
```
nit: typing

robertgshaw2-redhat commented:
LGTM. Thanks. Do you know if we have any test coverage for this metric?

markmc committed:
[V1][Metrics] Fix type hints

PR feedback from Rob.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
> LGTM, I like the abstraction and lifecycle. Just one nit on the typing. Ping when ready for automerge.

Thanks! Fixed

> Do you know if we have any test coverage for this metric?

No we don't, and I did consider addressing that (instead of manually) testing, but I was loathe to invest more time in it because I'm only adding it for compatibility reasons ... I'm pretty dubious that it's the right long-term solution (so I'd rather e.g. spend that time capturing all of the input on the topic in the design doc)

hidva commented on vllm/v1/engine/output_processor.py:
```diff
@@ -246,6 +260,7 @@ def _update_stats_from_finished(self, req_state: RequestState,
         iteration_stats.update_from_finished_request(finish_reason,
                                                      request_output,
                                                      req_state.stats)
+        self.lora_states.finish_request(req_state)
```
If `iteration_stats` is None, the execution of `self.lora_states.finish_request(req_state) `will be skipped. Could this lead to a lora_state leak?

robertgshaw2-redhat replied:
You are correct. This would cause a leak in the case that:
- stat logging is disabled
- lora is enabled

Thanks for your detailed review!
